async-nats/tests/client_tests.rs

```diff
@@ -618,6 +618,92 @@ mod client {
         rx.recv().await;
     }
 
+    #[tokio::test]
+    async fn reconnect_delay_callback_custom() {
+        let server = nats_server::run_basic_server();
+
+        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let _ = ConnectOptions::new()
+            .retry_on_initial_connect()
+            .reconnect_delay_callback(move |attempts| {
+                let tx = tx.clone();
+
+                let duration =
+                    std::time::Duration::from_millis(std::cmp::min((attempts * 10) as u64, 100));
+                // report back the number of attempts
+                tx.send((attempts, duration)).unwrap();
+
+                duration
+            })
+            .connect(server.client_url())
+            .await
+            .unwrap();
+
+        drop(server);
+
+        let (attempt, duration) = rx.recv().await.unwrap();
+        assert_eq!(attempt, 1);
+        assert_eq!(duration.as_millis(), 10);
+
+        let (attempt, duration) = rx.recv().await.unwrap();
+        assert_eq!(attempt, 2);
+        assert_eq!(duration.as_millis(), 20);
+
+        let (attempt, duration) = rx.recv().await.unwrap();
+        assert_eq!(attempt, 3);
+        assert_eq!(duration.as_millis(), 30);
+
+        let (attempt, duration) = rx.recv().await.unwrap();
+        assert_eq!(attempt, 4);
+        assert_eq!(duration.as_millis(), 40);
+    }
+
+    #[tokio::test]
+    async fn reconnect_delay_callback_default() {
+        let server = nats_server::run_basic_server();
+        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+        let _client = ConnectOptions::new()
+            .event_callback(move |err| {
+                let tx = tx.clone();
+                async move {
+                    tx.send(err.to_string()).unwrap();
+                }
+            })
+            .connect(server.client_url())
+            .await
+            .unwrap();
+        drop(server);
+
+        let duration = std::time::Instant::now();
+        rx.recv().await;
+        rx.recv().await;
+
+        // Ensure that default backoff works as expected
+        let elapsed = duration.elapsed().as_millis();
+        assert_eq!(elapsed, 0);
```

Comment:
This also will be flake'y.

I would just check if first few reconnect happens in short interval, and then if it stabilizes at our max at one or two checks. Again, you need to take CI muddy behaviour into account.